kloud/provider: Add SoftLayer Support

acbodine opened the pull request:
Adds SoftLayer support as a kloud provider, and necessary frontend changes.

## Description
Following the example here [https://github.com/koding/kloud-provider-example](https://github.com/koding/kloud-provider-example) and the PRs for [DigitalOcean](https://github.com/koding/koding/pull/9187) and [Google](https://github.com/koding/koding/pull/9185) support, this PR include the necessary changes to enable SoftLayer provider support for kloud.

Part of this PR includes updates to the vendor Terraform source in this repository, and essentially consists of getting the SoftLayer provider directory from Terraform's master branch and dropping it in place here. The current SoftLayer support that Terraform has is basically provisioning virtual guests (or vms) and managing ssh keys.

This PR also includes most (maybe all?) of the UI/frontend changes necessary for a user to use Koding with SoftLayer.

## Motivation and Context
Implement a solution to #9073.

## How Has This Been Tested?
- Unit tests to validate basic functionality for `stack` `machine` and `schema` definitions
- Integration tests that rely on environment variables to actually call SoftLayer apis where necessary in validating `stack` and `machine` definitions
- I found mentions on Slack channel about the `downtowork` koding development deployment, so I created a development environment [http://downtowork-acbodine.koding.team:8090](http://downtowork-acbodine.koding.team:8090). In order to test my changes with the whole system I simply pulled my changes into this enviroment and rebuilt/re-ran the necessary Docker containers. You can create a team there, or use the one I have been developing with [http://dragons.downtowork-acbodine.koding.team:8090](http://dragons.downtowork-acbodine.koding.team:8090) and you can leverage the SoftLayer stack I have in that team `Voltorb SoftLayer Stack`
- The frontend/user testing that I did has been manual so far in order to verify that I could
  1. Verify Credentials user enters
  2. Save SoftLayer stack and generate Bootstrap resources, verified in SoftLayer UI
  3. Initialize SoftLayer virtual guest, and verified connectivity back to Koding client/UI
  4. Tested Koding UI displaying accurate virtual guest state info, as well as stop/start actions

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

acbodine committed:
Adds beginnings of Softlayer Kloud provider implementation

acbodine committed:
Adds softlayer provider for terraform

acbodine committed:
Implements ApplyTemplate for Softlayer stacks

acbodine committed:
Cleaning up and implementing Machine methods for SoftLayer provider

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/koding/koding?pullRequest=9543) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/koding/koding?pullRequest=9543) before we can accept your contribution.<br/>

gokmen commented on client/home/lib/virtualmachines/components/machineslist/machinedetails.coffee:
```diff
@@ -307,7 +307,7 @@ generateSpecs = (machine) ->
   { instanceTypes } = globals.config.providers[providerName]
 
-  instanceType = jMachine.meta?.instance_type ? configs[providerName]['base-vm']
+  instanceType = jMachine.meta?.instance_type ? globals.config.providers[providerName]['base-vm']
```
👍

gokmen reviewed (approved):
LGTM on node/client side 👍

rjeczalik commented on go/src/koding/kites/kloud/provider/softlayer/stack_test.go:
```diff
@@ -0,0 +1,125 @@
+	}
+}
+
+func TestApplyTemplate(t *testing.T) {
```
Could you add a test that verifies the template before and after applying?

Like [this](https://github.com/koding/koding/blob/c38a33244c85d6721ef8796d051e2f8b9004e496/go/src/koding/kites/kloud/provider/marathon/stack_test.go#L28) or [this](https://github.com/koding/koding/blob/c38a33244c85d6721ef8796d051e2f8b9004e496/go/src/koding/kites/kloud/provider/azure/stack_test.go#L89) one.

acbodine replied:
I'll get this test added in asap.

rjeczalik replied:
You can copy `stripNondeterministicResources` and `equal` to your package and adjust them to your needs.

If you created `stack/provider/providertest` package and move them there to have only one copy of them, it would be really awesome.

rjeczalik replied:
(in the meantime I've added `object.ReplaceFunc` which does more-or-less what `stripNondeterministicResources` does but in more generic manner - it may be useful here)

rjeczalik commented on go/src/koding/kites/kloud/provider/softlayer/stack.go:
```diff
@@ -0,0 +1,213 @@
+		return nil, err
+	}
+
+	// NOTE: Should we shadow out private vars here?
```
Not needed. Shadowing is there to prevent leaking delagated credentials.

If you have an admin and a user within your team, and user should now know Softlayer credentials used to create resources, then admin creates a credential and shares it with the user. Since the template is augemented server-side, user never sees the real credentials used.

The shadowing is used to prevent malicious user to extract credentials from template, e.g. by echoing them to a file during vm creating.

Since user has no access currently to bootstrap templates, no need for shadowing.

acbodine replied:
Thanks for clearing that up, I'll remove the `NOTE`

rjeczalik commented on go/src/koding/kites/kloud/provider/softlayer/stack.go:
```diff
@@ -0,0 +1,213 @@
+
+	t.Resource["softlayer_virtual_guest"] = resource.VirtualGuest
+
+	// Don't show secret things in frontend logs to the user
```
In case of Softlayer probably it's not that important. The softlayer-go library misuses Basic Auth, instead of adding a header to a HTTP request it adds auth info to the hostname:

https://github.com/maximilien/softlayer-go/blob/master/client/http_client.go#L69

So if request times out, go's `net/http` library is logging the whole hostname with `*net.OpError` errors. So, until the lib is fixed, the credentials will leak. Just FYI.

acbodine replied:
I sort of know the guy who maintains the `softlayer-go` repo, I'll see if I can reach out to him to get this fixed. I'll leave the shadowing in place for now.

rjeczalik commented on go/src/koding/kites/kloud/provider/softlayer/stack.go:
```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		return err
+	}
+	_, err = service.GetObject(1234567)
```
Is `1234567` always a valid object in Softlayer?

acbodine replied:
Correct

rjeczalik commented on go/src/koding/kites/kloud/provider/softlayer/stack.go:
```diff
@@ -0,0 +1,213 @@
+	guest["ssh_keys"] = keys
+
+	guest["image"] = "UBUNTU_14_64"
+	// NOTE: Should we allow for user override here? Do all images work with klient?
```
Bootstrap templates are fully controlled by kloud currently (server-side), user can't edit it now.

We're working actively to make klient support more distros, just FYI.

rjeczalik reviewed (approved):
Looks good, just it would be ideal to have blanket test for ApplyTemplate as described above. But no crucial.

acbodine commented on go/src/koding/kites/kloud/provider/softlayer/softlayer_test.go:
```diff
@@ -0,0 +1,3 @@
+package softlayer_test
+
+import ()
```
Good catch, deleting this.

acbodine committed:
Adds more comprehensive ApplyTemplate test and moves common logic to shared providertest package for other kloud providers.

acbodine commented:
@rjeczalik I added the blanket test for ApplyTemplate and refactored the template verification logic you pointed me at to `stack/provider/providertest` package. Let me know if anything else needs to be done.

gokmen commented:
@acbodine please ignore DO-NOT-MERGE label it's related with deployment process, we'll merge this in once it's ready after the next deploy, fyi.

rjeczalik commented on go/src/koding/kites/kloud/provider/softlayer/stack.go:
```diff
@@ -0,0 +1,210 @@
+func (s *Stack) BootstrapTemplates(c *stack.Credential) ([]*stack.Template, error) {
+	vs := &bootstrapVars{
+		KeyName: fmt.Sprintf(
+			"koding-%s",
```
Please include team name also here:

```
fmt.Sprintf("koding-%s-%s", s.Arg(*stack.BootstrapRequest).GroupName, c.Identifier)
```

rjeczalik commented on go/src/koding/kites/kloud/provider/softlayer/stack_test.go:
```diff
@@ -0,0 +1,152 @@
+	}{
+		"single guest stack": {
+			"testdata/single-guest.json",
+			"testdata/single-guest.golden.json",
```
The file name should be `single-guest.json.golden`.

rjeczalik commented on go/src/koding/kites/kloud/stack/provider/providertest/apply.go:
```diff
@@ -0,0 +1,63 @@
+package providertest
```
Please gofmt this file.

rjeczalik commented:
@acbodine Looks good, just minor comments.

acbodine committed:
Minor fixes to formatting and names of ssh keys and test templates

acbodine commented:
@rjeczalik Those should be all set now.

rjeczalik commented:
@acbodine Thanks, this is really good one! 💥

rjeczalik commented:
@acbodine Thanks, this is really a good one! 💥

acbodine commented:
@gokmen I resolved the merge conflicts

gokmen commented:
thanks @acbodine we're waiting for softlayer test accounts, after tests @rjeczalik will merge this one 👍

rjeczalik commented:
Credential verification would need to be fixed, right now it's a false positive and during bootstrap we are getting `Access denied` error during creating SSH keys.

We can merge this and iterate on fixes in separate PRs.

/cc @gokmen @cihangir

rjeczalik commented:
@acbodine Thanks for Softlayer, really nice work!
